server/players.py

```diff
@@ -134,37 +134,32 @@ def write_message(self, message: dict) -> None:
         with suppress(DisconnectedError):
             self.lobby_connection.write(message)
 
-    def to_dict(self):
+    def to_dict(self) -> dict:
         """
         Return a dictionary representing this player object
         """
-
-        def filter_none(t):
-            _, v = t
-            return v is not None
-
-        return dict(
-            filter(
-                filter_none, (
-                    ("id", self.id),
-                    ("login", self.login),
-                    ("avatar", self.avatar),
-                    ("country", self.country),
-                    ("clan", self.clan),
-                    ("ratings", {
-                        rating_type: {
-                            "rating": self.ratings[rating_type],
-                            "number_of_games": self.game_count[rating_type]
-                        }
-                        for rating_type in self.ratings
-                    }),
-                    # Deprecated
-                    ("global_rating", self.ratings[RatingType.GLOBAL]),
-                    ("ladder_rating", self.ratings[RatingType.LADDER_1V1]),
-                    ("number_of_games", self.game_count[RatingType.GLOBAL]),
-                )
-            )
-        )
+        assert self.state is not None and self.state.value is not None
+
+        cmd = {
+            "id": self.id,
+            "login": self.login,
+            "avatar": self.avatar,
+            "country": self.country,
+            "clan": self.clan,
+            "state": self.state.value if self.lobby_connection else "offline",
```

Comment:
This seems a little counterintuitive as to what the expected value of self.state is when the  player is offline. Might it not be better to add a playerState offline?

Reply:
What is counterintuitive about it? If the player is offline then the state is "offline". If I add an "OFFLINE" state to the enum then I have to make sure that's kept in sync with the value of lobby_connection which complicates things and introduces the possibility of weird bugs.

Reply:
Couldn't you just do it the line after you delete the lobby_connection? How does it complicate things more than that?
And it would also be sufficient I think to just at least add the offline to the enum. As I had expected that to define all possible states a player could be in but then you find out there is another state hidden in the code. So it makes it harder to be sure what the allowable values are

Reply:
Maybe, but I really don't want any other places in the code suddenly getting a player state of 'OFFLINE', since that isn't really a state... Internally you are offline when your LobbyConnection is gone which will cause an exception to be thrown when the server tries to send a message to you, thats how the code expects to be signaled. But I needed something to signal to the client that the player isn't connected anymore, and putting that in state seemed to make sense.

Reply:
Its the same with the game_info message. The states that we send to the client aren't actually the same as the internal states.

Reply:
Maybe for now it would be better to simplify things and just have 'online' and 'offline'. Unless you are actually planning on making use of the other states in the client. That's probably safer too because otherwise the internal state tracking becomes kindof part of the server protocol, which makes changing it a lot more annoying.

Reply:
Well with the addition of the states I would probably switch to using those to determine things like is a player available to join a queue or host or join a game as I would be trusting those are more consistent than perusing the games list in the client because the client may not necessarily even see all the game messages